### PR TITLE
fix: propagate context in koa tracing

### LINKED ISSUE #593: koa async await child spans not working
We just activated this in our project and we have 1 root span per request and no child spans in our trace... resulting in 1 span for an entire request.

I did some debugging to attempt to understand the problem by using the `traceApi` to create a child span in our first koa middleware. The result is always null (this is the same result the `pg` plugin gives). Enabling the logLevel 4, I see:
* `WARN:@google-cloud/trace-agent: Custom Span API: Attempted to create child span without root`
* `WARN:@google-cloud/trace-agent: pg: Attempted to create child span without root`

We **do** use native `async/await` and no babel transpiling. I am running with `GCLOUD_TRACE_NEW_CONTEXT=1`. I confirmed the `cls-ah.js` module is being loaded (indicating the `GCLOUD_TRACE_NEW_CONTEXT=1`) is working.

It seems the `async_hooks` stuff is not working in our case.

Our current setup:
* `node v8.9.1` (also tested with v8.9.1)
* `koa@v2.3.4`
* `@google-cloud/trace-agent@2.3.1`
* `pg@6.4.2`
* __no babel, only native js__

Any ideas what else might be wrong? What to try???

### PR DESCRIPTION
**Pull request by kjin**
Fixes #593

The Koa tracing plugin ignored the return value of `api.wrap`, resulting in trace context not being propagated correctly. This affects both Koa 1 and 2, but Koa 1 needs different patching logic altogether (as it relies on generator functions).

### CODE REVIEW

File: src/plugins/plugin-koa.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@
 'use strict';
 
 const shimmer = require('shimmer');
+const co = require('co');
 var urlParse = require('url').parse;
 
 function startSpanForRequest(api, req, res, next) {
```
Comment (jinwoo) [inline src/plugins/plugin-koa.ts]: It'd be good to declare the return type explicitly for a better readability.

File: src/plugins/plugin-koa.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@
 'use strict';
 
 const shimmer = require('shimmer');
+const co = require('co');
```
Comment (jinwoo) [inline src/plugins/plugin-koa.ts]: Why is this needed?

File: test/plugins/test-trace-koa.ts [hunk 3]
```diff
@@ -42,7 +42,7 @@ describe('koa', function() {
   });
 
   Object.keys(appBuilders).forEach(function(version) {
-    describe(version, function() {
+    describe.skip(version, function() {
```
Comment (jinwoo) [inline test/plugins/test-trace-koa.ts]: Why skip?

File: src/plugins/plugin-koa.ts [hunk 4]
```diff
@@ -94,9 +101,9 @@ function createMiddleware2x(api) {
   return function middleware(ctx, next) {
     const req = ctx.req;
     const res = ctx.res;
-
-    startSpanForRequest(api, req, res, next);
-
+    console.log(next);
```
Comment (jinwoo) [inline src/plugins/plugin-koa.ts]: Delete?

File: src/plugins/plugin-koa.ts [hunk 5]
```diff
@@ -94,9 +101,9 @@ function createMiddleware2x(api) {
...
+    next = startSpanForRequest(api, req, res, next);
+    console.log(next);
```
Comment (jinwoo) [inline src/plugins/plugin-koa.ts]: Delete?

File: src/plugins/plugin-koa.ts (hunk 1 above)
Comment (jinwoo) [reply src/plugins/plugin-koa.ts]: Can you put some type declarations here again?

File: src/plugins/plugin-koa.ts (hunk 5 above)
Comment (kjin) [reply src/plugins/plugin-koa.ts]: 😓

File: src/plugins/plugin-koa.ts (hunk 4 above)
Comment (kjin) [reply src/plugins/plugin-koa.ts]: Done

**Comment by kjin** [comment]
@jinwoo Yep, adding types for plugins is in the pipeline. I'm trying to figure out how to import types for multiple versions of the same module, which is the major blocker.

**Pull request merged**